docs: Sustainability effort (Incentive system) (#12359)
argoproj/argo-workflows: community/sustainability_effort.md

# Argo Workflows Sustainability Effort

Argo Workflows is seeking more community involvement and ultimately more [Reviewers and Approvers](https://github.com/argoproj/argoproj/blob/main/community/membership.md) to help keep it viable. 

We are trying out a system in which users pledge a certain number of hours per week (average) and in return have their own PRs prioritized for review.

Help is also needed for triaging new incoming bugs (prioritizing them with `P1`, `P2`, and `P3` labels).

> ## Terms

Reviewer: Did you mean "commitment"?

- Approver/Lead: should average a minimum of 2 hours per week of either PR review time or triaging bugs

Note that the hours per week from above is an average over time, and it's fine to have weeks of no activity.

Current roles for individuals can be found in [this document](https://github.com/argoproj/argoproj/blob/main/MAINTAINERS.md).

Participants should join the [#argo-wf-contributors](https://cloud-native.slack.com/archives/C0510EUH90V) slack channel.

### Finding code to work on

If you have a business need, definitely feel free to work on that. Otherwise, you can find an Issue and assign it to yourself. If you're a new developer, one option is to pick one up that has the 
label "good-first-issue". Otherwise, you can pick up an Issue labeled "P1" or "P2". These are the Issues deemed most important to the project.

### Finding a PR to review

Ideally, look for the Pull Requests which have no existing Assignee. An Assignee is a Reviewer/Approver/Lead who owns reviewing the PR. In the case of a Reviewer being the Assignee, the Reviewer 
can request a review from one or more Approvers once they have approved the PR.

### Bug Triaging

> We need to make sure that all new bugs are seen by somebody so that we can identify the highest priority ones. There are labels "P1", "P2", and "P3" that should be applied, in which "P1" is considered

Reviewer: Should also clarify what other priorities mean?

Author: done. Also realized that we do have a "P0" label as well.

highest priority. The "bug" label can be removed if the issue is determined to be user error. The label "more-information-needed" can be added if more information is needed from the author to in order
to determine whether it's a bug or its priority. If there's a new P1 bug which appears to be a showstopper, notify the [#argo-wf-contributors](https://cloud-native.slack.com/archives/C0510EUH90V) slack channel.

Any bugs that have >= 5 "thumbs up" reactions should be labeled `P1`. Any bugs with 3-4 "thumbs up" should be labeled `P2`. (Bugs can be sorted by "thumbs up").

## Participants

(if you'd like to join, just add your name here and submit in a PR)

| Name                      | GitHub ID                                               |
|---------------------------|---------------------------------------------------------|
| Julie Vogelman            | [juliev0](https://github.com/juliev0)                   |

> | Saravanan Balasubramanian | [sarabala1979](https://github.com/sarabala1979)         |

Reviewer: Feel free to add me here :-)

Author: thank you!

# Argo Workflows Sustainability Effort

Argo Workflows is seeking more community involvement and ultimately more [Reviewers and Approvers](https://github.com/argoproj/argoproj/blob/main/community/membership.md) to help keep it viable. 

We are trying an Incentive System in which contributors pledge a certain number of hours per week (average) to the project and in return have their PRs prioritized for review.

## Where is help needed?

Help is needed for:
* [reviewing PRs](#reviewing-PRs)

> * [triaging](../docs/CONTRIBUTING.md#triaging-bugs) new incoming bugs by prioritizing them with `P0`, `P1`, `P2`, and `P3` labels

Reviewer 2: ```suggestion
* [triaging](../docs/CONTRIBUTING.md#triaging-bugs) new bugs by prioritizing them with `P0`, `P1`, `P2`, and `P3` labels
```

- remove redundant word; simpler language [per k8s style guide](https://kubernetes.io/docs/contribute/style/style-guide/#use-simple-and-direct-language)

* responding to questions in [Github Discussions](https://github.com/argoproj/argo-workflows/discussions)
* responding to questions in [CNCF Slack](https://argoproj.github.io/community/join-slack) in the `#argo-workflows` and `#argo-wf-contributors` channels 

## Commitment

Argo Workflows has roles as described in the [Contributing](../docs/CONTRIBUTING.md#roles) document.
Any code contributor in a designated role or with an open Membership request can sign up to participate in this Incentive System.

Participants are expected to try to advance in roles.
There are different expectations depending on the role:

> - Member and below: should average a minimum of 8 hours per week of either [authoring PRs](../docs/CONTRIBUTING.md#authoring-prs) or performing any activities listed above in the ["Where is help needed?"](#where-is-help-needed) section

Reviewer: "Member and non-members" or "maintainers"

Author: I am going to go with the former since "maintainers" is not a clear concept in the [membership guide](https://github.com/argoproj/argoproj/blob/main/community/membership.md)

Reviewer: https://github.com/argoproj/argoproj/blob/main/community/GOVERNANCE.md#roles-and-membership

Reviewer: >For the purpose of governance, reviewers, approvers and leads are considered maintainers. The list of current maintainers, their project and organizational affiliation(s)is maintained in the [MAINTAINERS](https://github.com/argoproj/argoproj/blob/master/MAINTAINERS.md) file.

Author: good to know, but I think this is more clear, so if you're okay with it, I'll just keep it as is

Reviewer: Sure

Reviewer: I did small modification to make this clearer. https://github.com/argoproj/argoproj/pull/268

  - Note that more time is required in this role in order to learn the codebase and processes to make sufficient progress
- Reviewer and above: should average a minimum of 2 hours per week performing any activities listed above in ["Where is help needed?"](#where-is-help-needed) section 

>   - Highest priority is to ensure that all [`prioritized` PRs](#authoring-prs) have an Assignee

Reviewer: "p0"

Author: per our slack discussion, since there are only two levels, I prefer not to use "P0" since it implies that there's a "P1", "P2", and "P3", which there aren't

Reviewer: I am ok with "prioritized"

Author: thanks. Hope it's okay - I went with "prioritized-review" since it helps distinguish these from our priority labels for bugs.

Reviewer 2: Linking [the Slack discussion](https://cloud-native.slack.com/archives/C0510EUH90V/p1704221439136509?thread_ts=1704221270.952099&cid=C0510EUH90V) referenced above for posterity

>   - Highest priority is to ensure that all PRs labeled ["prioritized-review"](#authoring-prs) have an Assignee

Reviewer 2: ```suggestion
  - Highest priority is to ensure that all PRs labeled `prioritized-review` [have an Assignee]((#reviewing-prs)
```

- Use `prioritized-review` with backticks instead of quotation, for consistency with the rest of the doc and all other label references
- Linked the "have an Assignee" text instead, as I would expect a link on a label to just go to that label, and not a section of the doc
- Linked to "Reviewing PRs" section as that is the one that has instructions on this

Note that the hours per week listed above is an average over time; it's fine to have weeks of no activity so long as the average meets expectations.

> Participants should join the [#argo-wf-contributors](https://cloud-native.slack.com/archives/C0510EUH90V) slack channel.

Reviewer 2: ```suggestion
Participants should join the [`#argo-wf-contributors`](https://cloud-native.slack.com/archives/C0510EUH90V) Slack channel.
```

For consistency, same as above

### Reviewing PRs

> First priority is to review any PRs which have the `prioritized-review` label, meaning they were authored by a participant of this system (verified by GitHub ID).

Reviewer 2: ```suggestion
First priority is to review any PRs which have the [`prioritized-review` label]([`prioritized-review` label](https://github.com/argoproj/argo-workflows/labels/prioritized-review), meaning they were authored by a participant of this system (verified by GitHub ID).
```

added a link to the label

Each of these should have an Assignee: a Reviewer, Approver, or Lead who "owns" reviewing the PR.<br />
These PRs should be given an initial review within a week.
The Assignee should then respond to each question or requested modification from the author within a week.

Next, look for PRs with no Assignee. 

> For general guidance regarding PR Reviews, see [Reviewing PRs](../CONTRIBUTING.md#reviewing-prs).

Reviewer 2: ```suggestion
For general guidance regarding PR Reviews, see [the Contributing docs](../CONTRIBUTING.md#reviewing-prs).
```

this section itself is named "Reviewing PRs" (in both docs) so that sounds like a self-reference without seeing that the link is to a different docs page. we can distinguish it by saying "Contributing docs" instead

### Authoring PRs

Participants can apply a `prioritized-review` label to any PRs they author.

## Participants

If you'd like to participate, add yourself here in a PR.

| Name                      | GitHub ID                                               |
|---------------------------|---------------------------------------------------------|
| Julie Vogelman            | [juliev0](https://github.com/juliev0)                   |
| Saravanan Balasubramanian | [sarabala1979](https://github.com/sarabala1979)         |
| Yuan Tang                 | [terrytangyuan](https://github.com/terrytangyuan)       |

> | Alan Clucas               | [Joibel](https://github.com/Joibel)                     |

Reviewer 2: Can add me now as I definitely meet (or exceed) the current expectations
```suggestion
| Alan Clucas               | [Joibel](https://github.com/Joibel)                     |
| Anton Gilgur              | [agilgur5](https://github.com/agilgur5)                |
```

Also we probably want to alphabetize these? Right now they are assorted?
And we may want to put GitHub IDs in backticks

Author: Thanks for joining the effort! I know of course you already put in much more than the required minimum hours.

Author: It's interesting that the lack of backticks doesn't trigger the spelling error. Perhaps there's an exception for links?

Reviewer 2: Yea I noticed that too, I think `mdspell` may not be looking at the `community/` directory at all and [only the `docs/` directory right now](https://github.com/argoproj/argo-workflows/blob/c71e1dc05a052778bed597f7c0f50b2ede4a7f32/Makefile#L687)

Author: I see

# Argo Workflows Sustainability Effort

Argo Workflows is seeking more community involvement and ultimately more [Reviewers and Approvers](https://github.com/argoproj/argoproj/blob/main/community/membership.md) to help keep it viable. 

We are trying an Incentive System in which contributors pledge a certain number of hours per week (average) to the project and in return have their PRs prioritized for review.

## Where is help needed?

> Help is needed for:

Author: Gosh, I'm not sure if this is necessarily the priority order, though. In fact, I think the priority order for an individual is based on how they can best help. And below it says the highest priority is to address the PRs with the "prioritized" label, as opposed to PRs in general. It may be sufficient to make that the only priority, and everything else is just considered valuable.?

> Help is needed for:

Reviewer 2: Oh. Well that's true I suppose (although usually implied). I suppose the priority order makes more sense specifically for folks who are in fact capable of doing all of the above, which are indeed generally Reviewer+

If we want to incentivize reviewing PRs in general (as the area with the most need), I do think PRs should still be the highest priority if possible, and `prioritized` PRs above that.

But yea I think we can leave this as is then and only mention that below for Reviewer+. Let's make that a sub-bullet as opposed to parenthesized in the below section then

> Help is needed for:

Author: will do once I'm on a device where I can

> * reviewing PRs

Reviewer 2: ```suggestion
* [reviewing PRs](#reviewing-PRs)
```

add a link to the below section. I suggested renaming it as well

* [triaging](../docs/CONTRIBUTING.md#triaging-bugs) new incoming bugs by prioritizing them with `P0`, `P1`, `P2`, and `P3` labels
* responding to questions in [Github Discussions](https://github.com/argoproj/argo-workflows/discussions)
* responding to questions in [CNCF Slack](https://argoproj.github.io/community/join-slack) on `#argo-workflows` and `#argo-wf-contributors` channels 

## Commitment

Argo Workflows has roles as described in the [Contributing](../docs/CONTRIBUTING.md#roles) document.
Any code contributor in a designated role or with an open Membership request can sign up to participate in this Incentive System.

> To create a Membership request, create an Issue in the [argoproj](https://github.com/argoproj/argoproj) repository as described in the [Membership](https://github.com/argoproj/argoproj/blob/main/community/membership.md) guide.

Reviewer 2: In the Contributing Guide, I think we'd reword it a bit

```suggestion
If you are interested in formally joining the Argo project, [create a Membership request](https://github.com/argoproj/argoproj/issues/new?template=membership.md&title=REQUEST%3A%20New%20membership%20for%20%3Cyour-GH-handle%3E) in the [argoproj](https://github.com/argoproj/argoproj) repository as described in the [Membership](https://github.com/argoproj/argoproj/blob/main/community/membership.md) guide.
```

- also added a direct link to the membership issue template

> To create a Membership request, create an Issue in the [argoproj](https://github.com/argoproj/argoproj) repository as described in the [Membership](https://github.com/argoproj/argoproj/blob/main/community/membership.md) guide.

Author: Sure. Where would you like me to add it in the Contributing guide?

> To create a Membership request, create an Issue in the [argoproj](https://github.com/argoproj/argoproj) repository as described in the [Membership](https://github.com/argoproj/argoproj/blob/main/community/membership.md) guide.

Reviewer 2: I think at the bottom of the "Roles" section would be fine

Participants are expected to try to advance in roles.
There are different expectations depending on the role:

> - Member and below: should average a minimum of 8 hours per week of either [authoring PRs](../docs/CONTRIBUTING.md#authoring-prs) or doing any of the activities listed above in ["Where is help needed?"](#where-is-help-needed) section (more time is required in this role in order to make sufficient progress)

Reviewer 2: ```suggestion
- Member and below: should average a minimum of 8 hours per week of either [authoring PRs](../docs/CONTRIBUTING.md#authoring-prs) or performing any activities listed above in the ["Where is help needed?"](#where-is-help-needed) section
  - Note that more time is required in this role in order to learn the codebase and processes to make sufficient progress
```

- "performing" is a bit more accurate & formal wording than "doing"
- slightly simplify first sentence (per [k8s style guide](https://kubernetes.io/docs/contribute/style/style-guide/#use-simple-and-direct-language))
- add missing article/"the"
- split parenthesized part into a second sentence / sub-bullet note
  - also describe a bit more _why_ more time is needed

> - Reviewer and above: should average a minimum of 2 hours per week doing any of the activities listed above in ["Where is help needed?"](#where-is-help-needed) section (highest priority is to ensure that all [`prioritized` PRs](#authoring-prs) have an Assignee)

Reviewer 2: ```suggestion
- Reviewer and above: should average a minimum of 2 hours per week performing any activities listed above in the ["Where is help needed?"](#where-is-help-needed) section
```
- same as above on "performing" and simplifying sentence
- I added a prioritization order in a suggestion to the "Where is help needed?" section, so I think we can remove the parenthesized part with that

Author: modified the first bullet, but there's an outstanding question on the second so I left it as is

Help is also needed for triaging new incoming bugs (prioritizing them with `P0`, `P1`, `P2`, and `P3` labels).

## Commitment

The Argo project currently has 4 designated [roles](https://github.com/argoproj/argoproj/blob/main/community/membership.md):
- Member
- Reviewer
- Approver
- Lead

For those who focus on just one area of the code (such as UI), the `Reviewer` and `Approver` roles can optionally be scoped to just that area.

Anybody in any of these roles can sign up to be a part of this Incentive System. Participants are expected to try to advance in roles. The minimum expectations are different depending on which role:
- Member: should focus primarily on authoring PRs, should average a minimum of 8 hours per week; can also assist with triaging bugs or optionally reviewing PRs
- Reviewer: should author PRs and review PRs to move into the "Approver" role, should average a minimum of 2 hours per week of either PR review time or triaging bugs
- Approver/Lead: should average a minimum of 2 hours per week of either PR review time or triaging bugs

Note that the hours per week from above is an average over time, and it's fine to have weeks of no activity.

> Current roles for individuals can be found in [this document](https://github.com/argoproj/argoproj/blob/main/MAINTAINERS.md).

Reviewer 2: Follow-up thread: https://github.com/argoproj/argo-workflows/pull/12359#discussion_r1430641277

### Finding code to work on

If you have a business need, definitely feel free to work on that. Otherwise, you can find an Issue and assign it to yourself. If you're a new developer, one option is to pick one up that has the 
label `good-first-issue`. Otherwise, you can pick up an Issue labeled `P1` or `P2`. 

### Finding a PR to review

We will first prioritize review of any PRs which have the "prioritized" label, meaning they were authored by a participant of this sytem (we can verify that they in fact are by github ID). We need to make 
sure each of these has an Assignee. An Assignee is a Reviewer/Approver/Lead who owns reviewing the PR. PRs with this label should be given an initial review within a week and should respond to each question 
or modification from the author within a week.

Next, look for the Pull Requests which have no existing Assignee. 

In the case of a Reviewer being the Assignee, once the Reviewer has approved the PR, they can request a review from one or more Approvers.

### Authoring PRs

For participants of the system, any PRs that you author you can apply a "prioritized" label to.

### Bug Triaging

> We need to make sure that all new bugs are seen by somebody so that we can identify the highest priority ones. There are labels `P0`, `P1`, `P2`, and `P3` that should be applied, in which `P0` is considered
> highest priority and needs immediate attention, followed by `P1`, `P2`, and then `P3`. The "bug" label can be removed if the issue is determined to be user error. The label `more-information-needed` can be added 
> if more information is needed from the author to in order to determine whether it's a bug or its priority. If there's a new `P0` bug, notify the [#argo-wf-contributors](https://cloud-native.slack.com/archives/C0510EUH90V) slack channel.
> 
> Any bugs that have >= 5 "thumbs up" reactions should be labeled `P1`. Any bugs with 3-4 "thumbs up" should be labeled `P2`. (Bugs can be sorted by "thumbs up").

Reviewer 2: ```suggestion
New bugs need to be triaged to identify the highest priority ones.

Apply the labels `P0`, `P1`, `P2`, and `P3`, where `P0` is highest priority and needs immediate attention, followed by `P1`, `P2`, and then `P3`.
If there's a new `P0` bug, notify the [#argo-wf-contributors](https://cloud-native.slack.com/archives/C0510EUH90V) slack channel.

Any bugs with >= 5 "👍" reactions should be labeled `P1`.
Any bugs with 3-4 "👍" reactions should be labeled `P2`. 
Bugs can be [sorted by "👍"](https://github.com/argoproj/argo-workflows/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc+label%3Abug).

If the issue is determined to be a user error and not a bug, remove the `bug` label (and the `regression` label, if applicable) and replace it with the `support` label.
If more information is needed from the author to diagnose the issue, then apply the `more-information-needed` label.
```

- Use active voice and present tense here where possible as these are directions, [per k8s style guide](https://kubernetes.io/docs/contribute/style/style-guide/#use-present-tense)
- Simpler and more direct language, [per k8s style guide](https://kubernetes.io/docs/contribute/style/style-guide/#use-simple-and-direct-language)
- Prefer 1 sentence per line of markdown [per style guide](https://argoproj.github.io/argo-workflows/doc-changes/)

- "thumbs up" -> "👍" to be more specific
- add a link to the sort
- consistently use `` backticks around label names
- replace "determine whether it's a bug or its a priority" with "diagnose the issue"

## Participants

(if you'd like to join, just add your name here and submit in a PR)

| Name                      | GitHub ID                                               |
|---------------------------|---------------------------------------------------------|
| Julie Vogelman            | [juliev0](https://github.com/juliev0)                   |
| Saravanan Balasubramanian | [sarabala1979](https://github.com/sarabala1979)         |
| Yuan Tang                 | [terrytangyuan](https://github.com/terrytangyuan)       |
| Alan Clucas               | [Joibel](https://github.com/Joibel)                     |
| Isitha Subasinghe         | [isubasinghe](https://github.com/isubasinghe)           |
| Jason Meridth             | [jmeridth](https://github.com/jmeridth)                 |

> | Shuangkun Tian            | [shuangkun](https://github.com/shuangkun)               |

Reviewer 3: > Per the current system as is listed here, Shuangkun would not be eligible as Argoproj Membership is a prerequisite

Can you give me some advices to be a member. Thanks! I am an owner of a product in AlibabaCloud named Managed Argo Workflow(like aws MWAA)， which helps user have a better use in mlops、large scale、cicd scene and we also have many requests from users in product. So I want to more participate in community work to help the product work well and also let my PRs quickly reviewed.

> | Shuangkun Tian            | [shuangkun](https://github.com/shuangkun)               |

Author: good catch @agilgur5. @shuangkun can you create an Issue in the argoproj repo?: https://github.com/argoproj/argoproj/issues similar to others you'll find in there that read "REQUEST: New membership for x"?

> | Shuangkun Tian            | [shuangkun](https://github.com/shuangkun)               |

Author: It looks like you've authored 10 PRs - I don't think there should be an issue with getting it merged, although need to find out if maybe we could do it prior to the usual membership meeting which only occurs once every 3 months.

> | Shuangkun Tian            | [shuangkun](https://github.com/shuangkun)               |

Reviewer 3: Thanks!  It has been merged 8. I'll try to get them all merged and continuous to submit prs.

> | Shuangkun Tian            | [shuangkun](https://github.com/shuangkun)               |

Reviewer 3: > Thanks! It has been merged 8. I'll try to get them all merged and continuous to submit prs.

https://github.com/argoproj/argoproj/issues/266    I have create a issue in the argoproj repo.

> | Shuangkun Tian            | [shuangkun](https://github.com/shuangkun)               |

Author: thanks @shuangkun ! I actually just spoke to @terrytangyuan, and he suggests we drop the requirement to be a member for this altogether. Maybe that makes the most sense since otherwise we run into this kind of problem. Are you good with that @agilgur5 ?

> | Shuangkun Tian            | [shuangkun](https://github.com/shuangkun)               |

Author: I have made this change just now, FYI

> | Shuangkun Tian            | [shuangkun](https://github.com/shuangkun)               |

Reviewer 2: I think we should at minimum require an open membership request, as that requires a non-trivial level of effort to meet the requirements for (compared to just adding yourself to this list without any contributions). We can call those participants as "pending members"

# Argo Workflows Sustainability Effort

Argo Workflows is seeking more community involvement and ultimately more [Reviewers and Approvers](https://github.com/argoproj/argoproj/blob/main/community/membership.md) to help keep it viable. 

We are trying an Incentive System in which contributors pledge a certain number of hours per week (average) and in return have their PRs prioritized for review.

Help is also needed for triaging new incoming bugs by prioritizing them with `P0`, `P1`, `P2`, and `P3` labels.

## Commitment

The Argo project currently has 4 designated [roles](https://github.com/argoproj/argoproj/blob/main/community/membership.md):
- Member
- Reviewer
- Approver
- Lead

> For those who focus on just one area of the code (such as UI), the `Reviewer` and `Approver` roles can optionally be scoped to just that area.

Reviewer 2: ```suggestion
The Reviewer and Approver roles can optionally be scoped to an area of the codebase (for example, UI or docs).
```

- more consistent sentence structure with the rest of the doc and simplified a bit [per k8s style guide](https://kubernetes.io/docs/contribute/style/style-guide/#use-simple-and-direct-language)
- the roles do not have backticks elsewhere in the doc, so remove them
- "for example" is a bit easier to understand for foreign speakers than "such as"
  - also added another example of docs

> Any person in a designated role (or even those who are not yet Members) can sign up to participate in this Incentive System.

Author: That's good. I wonder if we should say anything here about how to request.?

> Any person in a designated role (or even those who are not yet Members) can sign up to participate in this Incentive System.

Reviewer 2: It is described in [the Membership page](https://github.com/argoproj/argoproj/blob/main/community/membership.md) that you linked above, so I think that suffices, but we could add a small note & link if desired

Participants are expected to try to advance in roles.
There are different expectations depending on the role:

> - Member plus those not yet Member: should focus primarily on authoring PRs, should average a minimum of 8 hours per week; can also assist with triaging bugs or optionally reviewing PRs

Reviewer 2: ```suggestion
- Member and below: should focus primarily on authoring PRs, averaging a minimum of 8 hours per week. Can also assist with triaging bugs or reviewing PRs.
```

- fix run-on sentence
- shortened to "Member and below" as we do regularly use language like "Approver+" already

> - Reviewer: should author PRs and review PRs to move into the "Approver" role, should average a minimum of 2 hours per week of either PR review time or triaging bugs

Reviewer 2: ```suggestion
- Reviewer: should author and review PRs, averaging a minimum of 2 hours per week of either PR review time or triaging bugs
```

- same structure as above
- It already says above that "participants are expected to try to advance in roles", so the Approver portion is redundant

Reviewer 2: To clarify, does this mean that a Reviewer+'s authorship time is not considered as part of the system, but only review and triage time is conisdered?

Reviewer 2: Also what about responding to Slack and GitHub Discussions? I spend a lot of time in those (although discussion volume has decreased recently while issues have increased) and those are considered for Membership requests.

Author: As for the first comment, I specifically added 'to move into the "Approver" role' as a clarification for why it was distinct from what "Approver" was doing. I think it could be useful, no?

Author: The Slack and Github Discussion thing is interesting - perhaps we should expand into that as well.

Reviewer 2: > As for the first comment, I specifically added 'to move into the "Approver" role' as a clarification for why it was distinct from what "Approver" was doing. I think it could be useful, no?

Since progressing through roles is the baseline expectation, that sounds redundant to me.

In terms of distinctions, I suppose _within_ the incentive system itself (this specific doc), the _expectations_ (this specific list) are not necessarily altogether different for Reviewers vs Approver+. The Membership [guidelines](https://github.com/argoproj/argoproj/blob/main/community/membership.md) do list some other expectations, if you wanted to include some of those here, but I'm not sure if they're necessarily relevant to the incentive system _itself_, which is currently just about prioritizing PRs and so code review is the biggest piece.

Also what are your thoughts on considering PR authorship for Reviewer+? Might have missed that [comment above](https://github.com/argoproj/argo-workflows/pull/12359#discussion_r1433402502) (in between the other two).
Both reviews and authorship are needed for maintenance, but we can certainly say that some activities (including Slack and GH Discussions) are out-of-scope of the incentive system.
To be specific, I don't necessarily have an opinion either way on non-review activities, just want to make sure we consider all possible contributions and whether they are in- or out-of-scope for this incentive system (we could also explicitly list what is out-of-scope as well)

Author: > Since progressing through roles is the baseline expectation, that sounds redundant to me.
Let's come back to this one since it's more trivial than the rest.

> In terms of distinctions, I suppose _within_ the incentive system itself (this specific doc), the _expectations_ (this specific list) are not necessarily altogether different for Reviewers vs Approver+. The Membership [guidelines](https://github.com/argoproj/argoproj/blob/main/community/membership.md) do list some other expectations, if you wanted to include some of those here, but I'm not sure if they're necessarily relevant to the incentive system _itself_, which is currently just about prioritizing PRs and so code review is the biggest piece.
> 
> Also what are your thoughts on considering PR authorship for Reviewer+? Might have missed that [comment above](https://github.com/argoproj/argo-workflows/pull/12359#discussion_r1433402502) (in between the other two). Both reviews and authorship are needed for maintenance, but we can certainly say that some activities (including Slack and GH Discussions) are out-of-scope of the incentive system. To be specific, I don't necessarily have an opinion either way on non-review activities, just want to make sure we consider all possible contributions and whether they are in- or out-of-scope for this incentive system (we could also explicitly list what is out-of-scope as well). It is, after all, entirely optional, so not all types of contributions have to necessarily be included

As far as PR authorship for Approvers+: since this person has arrived at "Approver" level, I think that's sufficient, and don't have any specific expectation for authoring PRs. As far as PR authorship for Reviewers: yes, I do see some inconsistency between what I indicated for "Reviewer" and what I indicated for "Member-". On the other hand, I didn't want to ask for too much, and wanted to focus on what I considered to be important. So, I think I'd like to maybe just keep that as is.? 

As far as the Github Discussions and Slack Discussions, I guess we are seeing that unfortunately there's a mountain of work to be done. :( Anything we think is important for the health of the repo should probably be included. For slack I assume we are referring to `argo-workflows` channel and `argo-wf-contributors` channel? And I assume that generally speaking it would be Reviewer+ responding on `argo-wf-contributors` channel, but anyone responding on `argo-workflows` channel and Github Discussions?

Perhaps I can include all of these activities then, but make it so that if you're a "Reviewer+", the most important thing is that we address the "prioritized" PRs, making sure that those have an assignee.

Speaking of "prioritized" label: maybe will rename that to "prioritized-pr" so nobody accidentally applies it to an Issue? Or if you have a preference of something else, let me know.

Author: I have made all of these changes now (except for renaming the label).

Reviewer 2: > So, I think I'd like to maybe just keep that as is.?

I don't really have much of an opinion, so personally, I think it's fine either way.

> As far as the Github Discussions and Slack Discussions, I guess we are seeing that unfortunately there's a mountain of work to be done. :(

Yea I find support to be one of the biggest & oft unexpected parts of maintaining something.
_Personally_, in my own repos, I disable Discussions and have no messaging forum. The docs should be primary self-help resource and the focus should be on improvements to the docs rather than on support. Discussions can be a bit distracting, as you'll get support questions in both Issues and Discussions, except only one of them has issue templates etc 😕 
But CNCF strongly prefers having both.
And also support questions will come in one way or the other, it's more of a question of where. Issues are much more accessible to search than Slack or Discussions, so if you're gonna get support questions, might as well put them there. If GH had a more nuanced implementation of Discussions and better tooling for them, that would be ideal but 😕 

That's just my opinion / thought process, for reference. Argo certainly gets more support questions as it's fairly complicated with lots of configurability and also requires a k8s cluster (and also k8s familiarity) to get started, so folks tend to have more upfront questions as a result. That's my observation so far at least.

> Anything we think is important for the health of the repo should probably be included.

I do think support is important, but I also don't like users relying on free support and the dynamic that creates -- i.e. high expectations of free volunteer help.
Vendors with support contracts I think should be more of the norm, but as a whole it is part of one of the current debates in OSS that does not really have a good solution or standard in place. Users just getting started or exploring won't want to make a support contract just yet; although PoCs & pilots also exist for that purpose in the sales process.

The reason I spend more time on those was because:
- initially, knowledge as a user is how you can contribute best
  - linking to docs and contributing to docs after answering questions was another area I focused on and that I think is good for initial contributions
- it's a lot easier to participate there, especially from my phone, than on PRs and code reviews etc.

If we include those activities, then I can add my name as I already spend a sizeable amount of time there even during months I'm less active

> For slack I assume we are referring to `argo-workflows` channel and `argo-wf-contributors` channel?

Yes mostly the former

> And I assume that generally speaking it would be Reviewer+ responding on `argo-wf-contributors` channel, but anyone responding on `argo-workflows` channel and Github Discussions?

More or less yes. I think Member+ or even interested contributors would fit `#argo-wf-contributors`. I certainly used it a good bit as a Member and we do have contributors without roles there too.

> Perhaps I can include all of these activities then, but make it so that if you're a "Reviewer+", the most important thing is that we address the "prioritized" PRs, making sure that those have an assignee.

If we do include those, then yes I'd agree (again I don't really have a strong opinion on including them or not).
I think that makes the most sense to ensure the main target is still encouraging PR review, where we need the most help.

> Speaking of "prioritized" label: maybe will rename that to "prioritized-pr" so nobody accidentally applies it to an Issue? Or if you have a preference of something else, let me know.

Agree.
My ideal might be `triaged/prioritized-pr` or some category; personally I'd rename all `P0`-`P3` into the same `triaged` category as well (`triaged/P3` etc). A different category name could be `priority/` or `meta/` or something as well. `P3` etc is not specified to be for bugs in its name either.
I use categories pretty heavily in [other repos I maintain](https://github.com/ezolenko/rollup-plugin-typescript2/labels) as most things can be organized across a few categories. Tbh primary reason I'd want Approver status at this point is so that I can edit issue titles (which helps searchability and accuracy), labels, etc on my own 😅 

But nbd either way at this point of what to name it and whether to have a category or not. Not super important to this doc how its named

> - Approver/Lead: should average a minimum of 2 hours per week of either PR review time or triaging bugs

Reviewer 2: ```suggestion
- Approver and above: should average a minimum of 2 hours per week of either PR review time or triaging bugs
```

- same "Approver+" language we use elsewhere

> Note that the hours per week from above is an average over time, and it's fine to have weeks of no activity.

Reviewer 2: ```suggestion
Note that the hours per week listed above is an average over time; it's fine to have weeks of no activity so long as the average meets expectations.
```

small clarification

> Current roles for individuals can be found in [OWNERS](../OWNERS).

Author: @agilgur5 are you saying you think there's a problem with not having the members listed anywhere? Is this when we're reviewing the PR to add an individual here and need to verify? I'm not sure what your method was, but it seems to me that the Argo icon shows up in an individual's github profile if they're a member? and one can find a Closed PR in the argoproj repo for their membership as well?

Author: Perhaps I can at least clarify that it's reviewers+ that are listed in that document. Let me do that.

Reviewer 2: For posterity, this is a follow-up thread to https://github.com/argoproj/argo-workflows/pull/12359#discussion_r1429416409

Yea I think a clarification would suffice. The term "roles" is defined above as including Members and this line used "roles" while not including Members

> Current roles for Reviewers, Approvers, and Leads can be found in [OWNERS](../OWNERS).

Reviewer 2: ```suggestion
Current roles for Reviewers and above can be found in [OWNERS](../OWNERS).
```

- same as above

Participants should join the [#argo-wf-contributors](https://cloud-native.slack.com/archives/C0510EUH90V) slack channel.

### Finding code to work on

If you have a business need, definitely feel free to work on that.
Otherwise, you can find an issue and assign it to yourself.
If you're a new developer, one option is to pick one up that has the label `good-first-issue`.
You can also try to pick up higher priority issues labeled `P1` or `P2`. 

### Finding a PR to review

First priority is to review any PRs which have the `prioritized` label, meaning they were authored by a participant of this system (verified by GitHub ID).
Each of these should have an Assignee: a Reviewer, Approver, or Lead who "owns" reviewing the PR.<br />
These PRs should be given an initial review within a week.
The Assignee should then respond to each question or modification from the author within a week.

Next, look for PRs with no Assignee. 

If a Reviewer is the Assignee, then once they have approved the PR, they should request a review from one or more Approvers.

### Authoring PRs

Participants can apply a `prioritized` label to any PRs they author.

> ### Triaging Bugs

Reviewer 2: I mentioned this [above](https://github.com/argoproj/argo-workflows/pull/12359#pullrequestreview-1785741940) (and Terry had a [related comment](https://github.com/argoproj/argo-workflows/pull/12359#issuecomment-1859541541)) but I do think this section and [part of](https://github.com/argoproj/argo-workflows/pull/12359#discussion_r1429427867) the "Finding code to work on" section should be moved to the Contributing docs, as they are not specific to the incentive system and are rather more general maintenance norms that we follow. 

The initial roles description in "Commitment" could be moved to the Contributing docs as well. We can then link to the Contributing docs here instead.

Author: Yes, I agree.
